### DISCUSSION TITLE
How do we deploy GitHub Action artifacts to on-prem servers?

### DISCUSSION THREAD

**Discussion by Rod-at-DOH** [community/community#55778] [Enterprise] [answered]
### Select Topic Area

Question

### Body

We're migrating all our source code out of an old TFS instance into a new GitHub Enterprise Organization. I've been learning how to do GH Actions to build artifacts. But soon we'll have to take the created artifacts and deploy them to servers on-prem. I don't even know where to begin learning how to do that. Every tutorial I've seen deploys to Azure, or AWS, etc. This isn't going to a cloud environment. I'm sure there must be some way, or maybe ways, of doing this. I'm asking that I be directed to where I can learn how to do this, please.

**Comment by airtower-luna** [comment]
You'll probably want to use an automation tool like [Ansible](https://docs.ansible.com/ansible/latest/index.html) (there are alternatives, Ansible is what I'm familiar with) and define in there how to deploy your software. Then call that when you need to deploy.

One important thing to consider: Are the servers you're deploying from reachable from the Actions runners you use? If not you can't deploy directly from Actions, you'd have to create an internal tool that retrieves the right artifacts and deploys them.

**Reply by Rod-at-DOH** [reply]
Thank you for your reply. The servers I need to deploy to are not reachable from GitHub Actions runners. And this is the crux of the matter. I tried doing this with an Azure DevOps Services self-hosted agent. Does GitHub have anything similar? If not, then how do I go about an internal tool to retrieve the artifacts and deploy them?

**Reply by airtower-luna** [reply, marked as answer]
Using [self-hosted runners](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners) is an option, so you could run one inside the right network and orchestrate the deployment from there.

What I'd strongly recommend against is running the self-hosted runner on the machine to deploy to, use a proper automation tool to do the actual deployment remotely from the right network instead.

> If not, then how do I go about an internal tool to retrieve the artifacts and deploy them?

Use [the API](https://docs.github.com/en/rest?apiVersion=2022-11-28) to find the right artifacts (or release assets, depending on how your builds provide them), and download them. Then automate the deployment internally from there.

**Reply by Rod-at-DOH** [reply]
Good advice, @airtower-luna. I believe I've found the correct [documentation for this](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners). At the moment I'm trying to get proof of concept going, to prove this is the right way to go.

**Reply by gokhan-uysal** [reply]
Also you can use tailscale to connect github runners to your on prem nodes. They also have an action to do it: [tailscale-join-action](https://github.com/tailscale/github-action?tab=readme-ov-file)

**Reply by Rod-at-DOH** [reply]
I'm not familiar with Tailscale. I don't know if we even have that in-house, I'll investigate.

**Reply by gokhan-uysal** [reply]
Give it a try it is amazing and worth to learn

**Reply by Rod-at-DOH** [reply]
My Deputy CISO says one word:

**NOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO**

**Reply by gokhan-uysal** [reply]
ahahahhahahahhahahah can we learn the reason :)

**Reply by Rod-at-DOH** [reply]
Nope.

**Comment by iSatishYadav** [comment]
If this is not already too late, TFS now Azure DevOps Server (still on-prem) supports all things Git. Why move to GitHub?
